api/storage service: validate table names

bhalevy opened the pull request:
This series fixes a couple issues around generating and handling of no_such_keyspace and no_such_column_family exceptions.

First, it removes std::throw_with_nested around their throw sites in the respective database::find_* functions.
Fixes #9753

And then, it introduces a `validate_tables` helper in api/storage_service.cc that generates a `bad_param_exception` in order to set the correct http response status if a non-existing table name is provided in the `cf` http request parameter.
Fixes #9754

The series also adds a test for the REST API under test/rest_api that verifies the storage_service enable/disable auto_compaction api and checks the error codes for non-existing keyspace or table.

Test: unit(dev)

bhalevy commented:
In v2:
- fixed review comments from #9746 draft
- separated find_uuid change
  - added internal error processing when failing to synchronously find the uuid returned by find_uuid(ks, cf)
- used check_rest_api in cql-pytest
- simplified rest_api test for storage_service by just checking the response status_cope in negative test cases

nyh commented on api/storage_service.cc:
```diff
@@ -76,6 +76,22 @@ static sstring validate_keyspace(http_context& ctx, const parameters& param) {
     throw bad_param_exception("Keyspace " + param["keyspace"] + " Does not exist");
 }
 
+static std::vector<sstring> validate_tables(const sstring& ks_name, http_context& ctx, const std::unordered_map<sstring, sstring>& query_params, sstring param_name) {
```
nitpick: I would call it split_and_validate_tables or even parse_tables since it also does the splitting, not just validation.

bhalevy replied:
ok

nyh reviewed (changes requested):
Looks good. I only have a few minor comments left.

bhalevy commented on test/cql-pytest/run.py:
```diff
@@ -260,6 +261,13 @@ def check_ssl_cql(ip):
+def check_rest_api(ip, port=10000):
+    resp = requests.get(f"http://{ip}:{port}/ui/")
+    if resp.status_code != requests.codes.ok:
+        raise NotYetUp
```
@nyh did you check that?
What exception did you get?

nyh replied:
No, I didn't check it (to check this you should probably remove the wait for cql and leave only the wait for rest api). I believe that requests.get throws a ConnectionError if the host is not available - it doesn't return an http status_code because there is none.

bhalevy replied:
you're right, will fix.

bhalevy commented on test/cql-pytest/run.py:
```diff
@@ -260,6 +261,13 @@ def check_ssl_cql(ip):
 
+## Test that the Scylla REST API is serving, for wait_for_services() below.
+def check_rest_api(ip, port=10000):
+    resp = requests.get(f"http://{ip}:{port}/ui/")
```
I'm not what we get for '/' when scylla is not yet up or not listening on the port.
If we get 404 only when it starts listening then it's ok to rely on.

nyh replied:
"/" and "/ui" are just two different inputs to the HTTP server - if it's not up, both will fail exactly the same.
It is theoretically possible that we bring up an HTTP server that doesn't know how to answer anything (returns 404 to everything) and only then register the handlers, and then testing "/" would not be enough. But we don't do this - the HTTP server only starts serving after being set up with the handlers. So testing "/" should be fine. And again, I think the point is to check that requests.get() didn't throw - not what error code it returned.

bhalevy commented on test/rest_api/conftest.py:
```diff
@@ -24,10 +24,20 @@
 # By default, tests run against a Scylla server listening
 # on localhost:9042 for CQL and localhost:10000 for the REST API.
 # Add the --host, --port, --ssl, or --api-port options to allow overiding these defaults.
+# Use the util.py library from ../cql-pytest:
```
bad rebase merge...

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -0,0 +1,66 @@
+        resp.raise_for_status()
+
+        resp = rest_api.send("POST", f"storage_service/auto_compaction/{test_keyspace}")
+        resp.raise_for_status()
```
Both points are correct.
But since the default auto-compaction state might theoretically change in the future maybe it is a good idea to create a temporary keyspace for the test (for each case since it's not a fixture anymore)

nyh replied:
Yes, I think it will be better. I'm tempted to commit this as-is and let you send a followup patch for this if you want.

bhalevy commented on test/cql-pytest/run.py:
```diff
@@ -261,12 +261,12 @@ def check_ssl_cql(ip):
     resp = requests.get(f"http://{ip}:{port}/ui/")
     if resp.status_code != requests.codes.ok:
         raise NotYetUp
-    # Any other exception may indicate a problem, and is passed to the caller.
```
it's gone in the next version

bhalevy commented:
In v3:
- Fixed check_rest_api to rely on the exception thrown when the serve is not up
  - probe the root url, don't care about http success/error status
- Fixed comment placement in run script
- test_storage_service
  - added "_keyspace" to first test name
  - added local `new_keyspace` function
  - create and drop a keyspace for each storage_service test
- renamed `validate_tables` to `parse_tables`
- rebased + retested
